rfcs/rfc-bld-20220623-1-build-failure-rca.md

```diff
@@ -0,0 +1,167 @@
+      {
+        "_class" : "io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImpl$EdgeImpl",
+        "id" : "778",
+        "type" : "STAGE"
+      }
+    ]
+  },
+```
+2. **Setup Stage:** After the parallel stage is the setup stage, setup stage is to setup build node, pull source codes from Github and install build dependencies.
+```
+  {
+    "_class" : "io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImpl",
+    "displayName" : "Setup",
+    "id" : "778",
+    "result" : "SUCCESS",
+    "type" : "STAGE",
+    "edges" : [
+      {
+        "_class" : "io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImpl$EdgeImpl",
+        "id" : "1112",
+        "type" : "STAGE"
+      }
+    ]
+  },
+```
+3. **Build Stage:** Build stage is the stage runs the actual build task.
+```
+{
+    "_class" : "io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImpl",
+    "displayName" : "release",
+    "id" : "1112",
+    "result" : "SUCCESS",
+    "type" : "STAGE",
+    "edges" : [
+      {
+        "_class" : "io.jenkins.blueocean.rest.impl.pipeline.PipelineNodeImpl$EdgeImpl",
+        "id" : "1718",
+        "type" : "STAGE"
+      }
+    ]
+  },
+```
+The RCA dispatcher lambda function uses these pipeline stage data to build a relationship graph of build platform/configuration, and then it triggers the RCA worker lambda function to do build failure RCA for failed build platform/configuration.
+
+RCA worker lambda function sends a GET request
+```
+https://{JENKINS_URL}/blue/rest/organizations/jenkins/pipelines/{JOB_NAME}/runs/{BUILD_NUMBER}/nodes/{ID}/log
+```
+to get the build log of a certain build platform/configuration.
+#### Build Failure RCA Patterns
+RCA worker lambda function scans the build log and search for log messages that match build failure RCA patterns.
+
+Current build RCA patterns from https://github.com/aws-lumberyard/build-failure-rca/tree/main/build-failure-rca/rca-patterns can be reused.
+
+#### Workflow
+![workflow](rfc-bld-20220623-1-build-failure-rca/workflow.png)
+
+1. A change in build failure RCA repo triggers the build failure RCA CDK deployment pipeline.
+2. The build failure RCA CDK deployment pipeline deploys build failure RCA CDK stack to AWS accounts that are using build failure RCA system.
+3. Change in O3DE repo or O3DE fork repo triggers the AR or Periodic builds.
+4. After each AR or Periodic build completes, it publishes build data to post build SNS topic.
+5. Post build SNS topic triggers the RCA dispatcher lambda function.
+6. RCA dispatcher lambda function retrieves AR/Periodic build data and dispatch the actual RCA work for each build platform/configuration to separate RCA worker lambda functions. 
+7. RCA worker lambda function retrieves build log of a certain build platform/configuration, it scans the build log and search for log messages that match build failure RCA patterns, once the RCA work is done, it uploads the build log to S3 and return RCA result to RCA dispatcher lambda function.
```

Comment:
In the event of the worker failing to fire for some extended period of time (maybe due to a region outage), what would happen? Would some sort of rate limiting occur as the function catches up with SNS notifications?

Reply:
We will have a retry mechanism and set a timeout waiting for the response from the worker.